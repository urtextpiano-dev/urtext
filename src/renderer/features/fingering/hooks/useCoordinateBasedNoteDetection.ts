/**
 * Coordinate-based note detection for fingering clicks
 * 
 * This is a workaround for the data-note-id injection issue.
 * Instead of relying on DOM attributes, we use the graphicalNoteMap
 * to find notes based on click coordinates.
 */

import { useCallback, useRef } from 'react';
import { selectNoteFromTies } from '../utils/selectNoteFromTies';
import { extractFirstDataNoteId } from '@/renderer/utils/noteUtils';

interface BoundingBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

interface GraphicalNoteWithBounds {
  noteId: string;
  graphicalNote: any;
  bounds?: BoundingBox;
}

export const useCoordinateBasedNoteDetection = (
  graphicalNoteMap: Map<string, any> | undefined,
  containerRef: React.RefObject<HTMLElement>
) => {
  // Cache bounds for performance
  const boundsCache = useRef<Map<string, GraphicalNoteWithBounds>>(new Map());
  
  // Update bounds cache when graphicalNoteMap changes
  const updateBoundsCache = useCallback(() => {
    if (!graphicalNoteMap || !containerRef.current) return;
    
    boundsCache.current.clear();
    
    // üîç DEBUG: Track chord detection
    const notesByPosition = new Map<string, Array<{ noteId: string, bounds: any }>>();
    
    for (const [noteId, graphicalNote] of graphicalNoteMap) {
      try {
        // PRIORITY 1: Use PositionAndShape for accurate individual note bounds
        if (graphicalNote?.PositionAndShape?.AbsolutePosition && 
            graphicalNote?.PositionAndShape?.Size) {
          const { AbsolutePosition: pos, Size: size } = graphicalNote.PositionAndShape;
          const containerRect = containerRef.current.getBoundingClientRect();
          
          // OSMD uses 10 pixels per staff unit at 100% zoom
          // Get current zoom from parent OSMD instance if available
          const osmdZoom = (window as any).osmd?.zoom || 1;
          const pixelScale = 10 * osmdZoom;
          
          // Convert OSMD coordinates to pixel coordinates
          // Fix NaN issue: Use size.width/height if available, fallback to typical notehead dimensions
          const bounds = {
            x: pos.x * pixelScale,
            y: pos.y * pixelScale,
            width: (size.width || size.x || 20) * pixelScale,  // Fallback to 20px typical notehead width
            height: (size.height || size.y || 15) * pixelScale  // Fallback to 15px typical notehead height
          };
          
          // Store bounds with high confidence
          boundsCache.current.set(noteId, {
            noteId,
            graphicalNote,
            bounds
          });
          
          // Track chord positions
          const yKey = Math.round(bounds.y);
          if (!notesByPosition.has(yKey.toString())) {
            notesByPosition.set(yKey.toString(), []);
          }
          notesByPosition.get(yKey.toString())!.push({ noteId, bounds });
          
          console.log(`‚úÖ PositionAndShape bounds for ${noteId}:`, bounds);
          continue; // Skip SVG fallback
        }
        
        // FALLBACK: Original SVG element method
        let svgElement: SVGElement | null = null;
        
        if (typeof graphicalNote.getSVGGElement === 'function') {
          svgElement = graphicalNote.getSVGGElement();
        } else if (typeof graphicalNote.getSVGElement === 'function') {
          svgElement = graphicalNote.getSVGElement();
        }
        
        if (svgElement) {
          const rect = svgElement.getBoundingClientRect();
          const containerRect = containerRef.current.getBoundingClientRect();
          
          const bounds = {
            x: rect.left - containerRect.left,
            y: rect.top - containerRect.top,
            width: rect.width,
            height: rect.height
          };
          
          // üîç DEBUG: Detailed SVG element analysis
          console.log(`üéµ DEBUG SVG ELEMENT - ${noteId}:`, {
            noteId,
            svgElement: {
              tagName: svgElement.tagName,
              className: svgElement.className?.baseVal || svgElement.className,
              id: svgElement.id,
              children: svgElement.children.length,
              childTypes: Array.from(svgElement.children).map(child => ({
                tag: child.tagName,
                className: child.className?.baseVal || child.className
              }))
            },
            rawRect: {
              left: rect.left,
              top: rect.top,
              right: rect.right,
              bottom: rect.bottom,
              width: rect.width,
              height: rect.height
            },
            relativeBounds: bounds
          });
          
          // Store bounds relative to container
          boundsCache.current.set(noteId, {
            noteId,
            graphicalNote,
            bounds
          });
          
          // üîç DEBUG: Group notes by Y position to detect chords
          const yKey = Math.round(bounds.y);
          if (!notesByPosition.has(yKey.toString())) {
            notesByPosition.set(yKey.toString(), []);
          }
          notesByPosition.get(yKey.toString())!.push({ noteId, bounds });
        }
      } catch (error) {
        // Silent fail for individual notes
      }
    }
    
    // üîç DEBUG: Report chord groups
    notesByPosition.forEach((notes, yPos) => {
      if (notes.length > 1) {
        console.log(`üéπ CHORD DETECTED at Y=${yPos}:`, {
          noteCount: notes.length,
          notes: notes.map(n => ({
            id: n.noteId,
            x: n.bounds.x,
            sameY: notes.every(note => Math.abs(note.bounds.y - n.bounds.y) < 1),
            sameBounds: notes.every(note => 
              Math.abs(note.bounds.x - n.bounds.x) < 1 &&
              Math.abs(note.bounds.y - n.bounds.y) < 1 &&
              Math.abs(note.bounds.width - n.bounds.width) < 1 &&
              Math.abs(note.bounds.height - n.bounds.height) < 1
            )
          }))
        });
      }
    });
    
    console.log(`üìç Coordinate detection: Cached bounds for ${boundsCache.current.size} notes`);
  }, [graphicalNoteMap, containerRef]);
  
  // Find note at given coordinates
  const findNoteAtCoordinates = useCallback((clientX: number, clientY: number): string | null => {
    if (!containerRef.current) return null;
    
    const containerRect = containerRef.current.getBoundingClientRect();
    const relativeX = clientX - containerRect.left;
    const relativeY = clientY - containerRect.top;
    
    // üîç DEBUG: Log click details
    console.log(`üéØ CLICK ANALYSIS - Click at (${clientX}, ${clientY}) -> Relative (${relativeX.toFixed(1)}, ${relativeY.toFixed(1)})`);
    
    // Add some padding for easier clicking
    const clickPadding = 5;
    
    // Collect all matching notes (fixes chord note overlap issue)
    const matchingNotes: Array<{ noteId: string; bounds: any; distance: number }> = [];
    
    for (const [noteId, noteData] of boundsCache.current) {
      const { bounds } = noteData;
      if (!bounds) continue;
      
      if (
        relativeX >= bounds.x - clickPadding &&
        relativeX <= bounds.x + bounds.width + clickPadding &&
        relativeY >= bounds.y - clickPadding &&
        relativeY <= bounds.y + bounds.height + clickPadding
      ) {
        // Calculate Euclidean distance from click to note center
        const noteCenterX = bounds.x + bounds.width / 2;
        const noteCenterY = bounds.y + bounds.height / 2;
        const distance = Math.pow(relativeX - noteCenterX, 2) + Math.pow(relativeY - noteCenterY, 2);
        
        // üîç DEBUG: Distance calculation breakdown
        console.log(`üìê DISTANCE CALC - ${noteId}:`, {
          bounds: bounds,
          center: { x: noteCenterX.toFixed(1), y: noteCenterY.toFixed(1) },
          clickToCenter: {
            xDiff: (relativeX - noteCenterX).toFixed(1),
            yDiff: (relativeY - noteCenterY).toFixed(1),
            xDistSq: Math.pow(relativeX - noteCenterX, 2).toFixed(1),
            yDistSq: Math.pow(relativeY - noteCenterY, 2).toFixed(1),
            totalDistSq: distance.toFixed(1),
            totalDist: Math.sqrt(distance).toFixed(1)
          }
        });
        
        matchingNotes.push({ noteId, bounds, distance });
      }
    }
    
    if (matchingNotes.length === 0) {
      console.log(`‚ùå NO NOTES FOUND - Click at (${relativeX.toFixed(1)}, ${relativeY.toFixed(1)}) with padding ${clickPadding}px`);
      return null;
    }
    
    // Defensive check - ensure all matching notes have valid structure
    const validNotes = matchingNotes.filter(note => note && note.noteId && note.bounds);
    if (validNotes.length === 0) {
      console.error('‚ùå NO VALID NOTES - All matching notes missing required properties');
      return null;
    }
    
    // Use only valid notes for further processing
    const processedNotes = validNotes;
    
    // üîç DEBUG: Show all matching notes before selection
    if (matchingNotes.length > 1) {
      console.log(`üéµ CHORD CLICK - Found ${matchingNotes.length} matching notes:`);
      matchingNotes.forEach((match, idx) => {
        console.log(`  ${idx + 1}. ${match.noteId} (distance=${Math.sqrt(match.distance).toFixed(1)}px, distSq=${match.distance.toFixed(1)}, y=${match.bounds.y.toFixed(1)}, x=${match.bounds.x.toFixed(1)})`);
      });
      
      // üîç DEBUG: Check if bounds are truly identical
      const firstBounds = matchingNotes[0].bounds;
      const allIdentical = matchingNotes.every(note => 
        Math.abs(note.bounds.x - firstBounds.x) < 0.1 &&
        Math.abs(note.bounds.y - firstBounds.y) < 0.1 &&
        Math.abs(note.bounds.width - firstBounds.width) < 0.1 &&
        Math.abs(note.bounds.height - firstBounds.height) < 0.1
      );
      console.log(`üîç BOUNDS ANALYSIS - All bounds identical: ${allIdentical}`);
      
      // üß™ NEW DEBUG: Test proposed solutions
      console.log(`üß™ TESTING CHORD SOLUTIONS:`);
      
      // Test 1: Extract MIDI values from note IDs
      const midiData = matchingNotes.map(note => {
        const midiMatch = note.noteId.match(/midi(\d+)$/);
        const midiValue = midiMatch ? parseInt(midiMatch[1]) : null;
        return { noteId: note.noteId, midiValue, bounds: note.bounds };
      });
      console.log(`  üéπ MIDI VALUES:`, midiData);
      
      // Test 2: Check if GraphicalNote objects are different
      const graphicalNotes = matchingNotes.map(note => boundsCache.current.get(note.noteId)?.graphicalNote);
      console.log(`  üéº GRAPHICAL NOTE OBJECTS SAME:`, graphicalNotes[0] === graphicalNotes[1]);
      
      // Test 3: Try to access potential notehead data
      if (graphicalNotes[0] && graphicalNotes[1]) {
        console.log(`  üìç GRAPHICAL NOTE 1:`, {
          hasVfNote: !!graphicalNotes[0].vfnote,
          hasSourceNote: !!graphicalNotes[0].sourceNote,
          hasPositionAndShape: !!graphicalNotes[0].PositionAndShape
        });
        console.log(`  üìç GRAPHICAL NOTE 2:`, {
          hasVfNote: !!graphicalNotes[1].vfnote,
          hasSourceNote: !!graphicalNotes[1].sourceNote,
          hasPositionAndShape: !!graphicalNotes[1].PositionAndShape
        });
        
        // Try accessing PositionAndShape if available
        if (graphicalNotes[0].PositionAndShape && graphicalNotes[1].PositionAndShape) {
          console.log(`  üéØ POSITION DATA 1:`, graphicalNotes[0].PositionAndShape.AbsolutePosition);
          console.log(`  üéØ POSITION DATA 2:`, graphicalNotes[1].PositionAndShape.AbsolutePosition);
        }
      }
      
      // Test 4: Simulate X-position tiebreaker
      const xDistances = matchingNotes.map(note => ({
        noteId: note.noteId,
        xDistance: Math.abs(relativeX - (note.bounds.x + note.bounds.width / 2))
      }));
      const xWinner = xDistances.sort((a, b) => a.xDistance - b.xDistance)[0];
      console.log(`  üéØ X-POSITION TIEBREAKER: ${xWinner.noteId} (distances: ${xDistances.map(d => d.xDistance.toFixed(1)).join(', ')})`);
      
      // Test 5: Simulate MIDI tiebreaker with Y-position
      if (midiData.every(m => m.midiValue !== null)) {
        const sortedByMidi = [...midiData].sort((a, b) => a.midiValue! - b.midiValue!);
        const avgY = midiData.reduce((sum, n) => sum + n.bounds.y, 0) / midiData.length;
        const selectHigherPitch = relativeY > avgY;
        const midiWinner = selectHigherPitch ? sortedByMidi[sortedByMidi.length - 1] : sortedByMidi[0];
        console.log(`  üéµ MIDI TIEBREAKER: ${midiWinner.noteId} (click Y=${relativeY.toFixed(1)}, avgY=${avgY.toFixed(1)}, selectHigher=${selectHigherPitch})`);
      }
    }
    
    // Use tiebreaker logic when distances are equal (chord notes)
    const minDistance = Math.min(...processedNotes.map(n => n.distance));
    const tiedNotes = processedNotes.filter(n => Math.abs(n.distance - minDistance) < 1.0);
    
    let selectedNote;
    if (tiedNotes.length > 1) {
      // Calculate relative Y position for pure function
      const avgY = tiedNotes.reduce((sum, n) => sum + n.bounds.y, 0) / tiedNotes.length;
      const relativePosition = (relativeY - avgY + 25) / 50; // Normalize to 0-1 range
      
      // Use pure function to select note (fixes stale closure issue)
      selectedNote = selectNoteFromTies(tiedNotes, relativePosition);
      
      if (selectedNote) {
        console.log(`üéµ CHORD TIEBREAKER: Selected ${selectedNote.noteId} (relativeY=${relativeY.toFixed(1)}, position=${relativePosition.toFixed(2)})`);
      }
    } else {
      // Regular distance-based selection
      selectedNote = processedNotes.sort((a, b) => a.distance - b.distance)[0];
    }
    
    // Final safety check with defensive fallback
    if (!selectedNote) {
      console.error('‚ùå SELECTION ERROR: selectedNote is undefined, using first processed note as fallback');
      selectedNote = processedNotes[0];
    }
    
    // Additional defensive check - ensure we have a valid noteId
    if (!selectedNote || !selectedNote.noteId) {
      console.error('‚ùå CRITICAL ERROR: selectedNote missing noteId, cannot proceed');
      return null;
    }
    
    console.log(`‚úÖ SELECTED NOTE: ${selectedNote.noteId}${processedNotes.length > 1 ? ` (chose from ${processedNotes.length} matches)` : ''}`);
    return selectedNote.noteId;
  }, [containerRef]);
  
  // Alternative: Find note using OSMD's internal hit testing
  const findNoteUsingOSMD = useCallback((event: MouseEvent): string | null => {
    if (!graphicalNoteMap || !containerRef.current) return null;
    
    // Get all potential note elements at click position
    const elementsAtPoint = document.elementsFromPoint(event.clientX, event.clientY);
    
    // FIRST: Check ALL elements at click point for data-note-id
    // Start from the topmost element (first in array) which is the actual clicked element
    for (const element of elementsAtPoint) {
      if (element instanceof Element && element.hasAttribute('data-note-id')) {
        const noteId = extractFirstDataNoteId(element.getAttribute('data-note-id'));
        if (noteId) {
          return noteId;
        }
      }
    }
    
    // SECOND: If no direct hit, try closest() as fallback
    for (const element of elementsAtPoint) {
      if (element instanceof Element) {
        const noteElement = element.closest('[data-note-id]');
        if (noteElement && noteElement.hasAttribute('data-note-id')) {
          const noteId = extractFirstDataNoteId(noteElement.getAttribute('data-note-id'));
          if (noteId) {
            return noteId;
          }
        }
      }
    }
    
    // FALLBACK: Removed - causes wrong chord note selection
    // The primary data-note-id lookup should handle all cases
    // If it fails, the click detection has deeper issues that need investigation
    
    return null;
  }, [graphicalNoteMap, containerRef]);
  
  return {
    updateBoundsCache,
    findNoteAtCoordinates,
    findNoteUsingOSMD
  };
};