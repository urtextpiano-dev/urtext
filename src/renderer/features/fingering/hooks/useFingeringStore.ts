/**
 * Re-export of fingering store for consistent imports
 * Phase 2 fingering-mvp-ui implementation
 */

export { useFingeringStore } from '../stores/fingeringStore';