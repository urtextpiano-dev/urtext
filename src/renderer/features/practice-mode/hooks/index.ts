/**
 * Practice Mode Hooks
 * 
 * Re-export all hooks for the practice mode feature
 */

export { usePracticeController } from './usePracticeController.v2';