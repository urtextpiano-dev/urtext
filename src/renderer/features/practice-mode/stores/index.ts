/**
 * Practice Mode Stores
 * 
 * Re-export all stores for the practice mode feature
 */

export { usePracticeStore } from './practiceStore';