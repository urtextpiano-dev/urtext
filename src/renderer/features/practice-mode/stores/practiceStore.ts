/**
 * Practice Mode Store
 * 
 * Zustand store for managing practice mode state.
 * Follows immutable update patterns and provides actions for state transitions.
 */

import { create } from 'zustand';
import { devtools } from 'zustand/middleware';
import type { PracticeState, PracticeStep, ComparisonResult, OptimizedPracticeStep } from '../types';

interface PracticeActions {
  /** Start a new practice session */
  startPractice: () => void;
  /** Stop the current practice session */
  stopPractice: () => void;
  /** Set the current step the user should play */
  setCurrentStep: (step: PracticeStep | null) => void;
  /** Update the set of currently pressed keys */
  updatePressedKeys: (keys: number[]) => void;
  /** Set the result of the last comparison */
  setResult: (result: ComparisonResult) => void;
  /** Update the current status */
  setStatus: (status: PracticeState['status']) => void;
  /** Reset attempt count for current step */
  resetAttempts: () => void;
  /** Set whether the score has repeats */
  setHasRepeats: (hasRepeats: boolean) => void;
  /** Dismiss the repeat warning */
  dismissRepeatWarning: () => void;
  /** Reset repeat warning dismissal (when new score loaded) */
  resetRepeatWarning: () => void;
  /** Toggle repeat navigation enabled/disabled */
  setRepeatsEnabled: (enabled: boolean) => void;
  /** Mark repeat navigation as failed */
  setRepeatsFailed: (failed: boolean) => void;
  
  // Phase 1 optimization: Pre-computed sequence management
  /** Set the pre-computed practice sequence */
  setOptimizedSequence: (sequence: OptimizedPracticeStep[]) => void;
  /** Get current optimized step by index */
  getCurrentOptimizedStep: () => OptimizedPracticeStep | null;
  /** Advance to next step in optimized sequence */
  advanceOptimizedStep: () => boolean;
  /** Reset optimized sequence position */
  resetOptimizedSequence: () => void;
  
  // V2 Controller sync
  /** Set isActive state directly (for V2 controller sync) */
  setIsActive: (isActive: boolean) => void;
  
  // Custom measure range actions
  /** Set custom measure range with validation */
  setCustomRange: (start: number, end: number) => void;
  /** Toggle custom range active state */
  toggleCustomRange: () => void;
  /** Clear custom range and reset to defaults */
  clearCustomRange: () => void;
  /** Set current measure index (0-based) */
  setCurrentMeasure: (measureIndex: number) => void;
  /** Set last valid measure index for recovery (0-based) */
  setLastValidMeasureIndex: (measureIndex: number) => void;
}

type PracticeStore = PracticeState & PracticeActions;

export const usePracticeStore = create<PracticeStore>()(
  devtools(
    (set) => ({
      // Initial state
      isActive: false,
      status: 'idle',
      currentStep: null,
      pressedKeys: new Set(),
      lastResult: null,
      attemptCount: 0,
      hasRepeats: false,
      repeatWarningDismissed: false,
      repeatsEnabled: true, // Default to ON for musical accuracy
      repeatsFailed: false, // Default to no errors
      
      // Phase 1 optimization: Pre-computed sequence state
      optimizedSequence: [] as OptimizedPracticeStep[],
      currentOptimizedIndex: 0,
      
      // Custom measure range state (with safe defaults)
      customRangeActive: false,
      customStartMeasure: 1,
      customEndMeasure: 1,
      currentMeasureIndex: undefined,
      lastValidMeasureIndex: 0,
      
      // Actions
      startPractice: () => set(
        { 
          isActive: true, 
          status: 'listening',
          attemptCount: 0,
          lastResult: null
        },
        false,
        'startPractice'
      ),
      
      stopPractice: () => set(
        { 
          isActive: false, 
          status: 'idle',
          currentStep: null,
          lastResult: null,
          pressedKeys: new Set(),
          attemptCount: 0,
          // Clear custom range when stopping practice
          customRangeActive: false,
          customStartMeasure: 1,
          customEndMeasure: 1
        },
        false,
        'stopPractice'
      ),
      
      setCurrentStep: (step) => set(
        { 
          currentStep: step,
          status: step ? 'listening' : 'idle',
          lastResult: null,
          attemptCount: 0
        },
        false,
        'setCurrentStep'
      ),
      
      updatePressedKeys: (keys) => set(
        { 
          pressedKeys: new Set(keys) 
        },
        false,
        'updatePressedKeys'
      ),
      
      setResult: (result) => set(
        (state) => ({ 
          lastResult: result,
          status: result.type === 'CORRECT' ? 'feedback_correct' : 'feedback_incorrect',
          attemptCount: state.attemptCount + 1
        }),
        false,
        'setResult'
      ),
      
      setStatus: (status) => set(
        { status },
        false,
        'setStatus'
      ),
      
      resetAttempts: () => set(
        { attemptCount: 0 },
        false,
        'resetAttempts'
      ),
      
      setHasRepeats: (hasRepeats) => set(
        { hasRepeats },
        false,
        'setHasRepeats'
      ),
      
      dismissRepeatWarning: () => set(
        { repeatWarningDismissed: true },
        false,
        'dismissRepeatWarning'
      ),
      
      resetRepeatWarning: () => set(
        { repeatWarningDismissed: false },
        false,
        'resetRepeatWarning'
      ),
      
      setRepeatsEnabled: (enabled) => set(
        { repeatsEnabled: enabled },
        false,
        'setRepeatsEnabled'
      ),
      
      setRepeatsFailed: (failed) => set(
        { repeatsFailed: failed },
        false,
        'setRepeatsFailed'
      ),
      
      // Phase 1 optimization: Pre-computed sequence actions
      setOptimizedSequence: (sequence) => set(
        { 
          optimizedSequence: sequence,
          currentOptimizedIndex: 0
        },
        false,
        'setOptimizedSequence'
      ),
      
      getCurrentOptimizedStep: () => {
        const state = usePracticeStore.getState();
        const { optimizedSequence, currentOptimizedIndex } = state;
        
        if (currentOptimizedIndex >= 0 && currentOptimizedIndex < optimizedSequence.length) {
          return optimizedSequence[currentOptimizedIndex];
        }
        return null;
      },
      
      advanceOptimizedStep: () => {
        const state = usePracticeStore.getState();
        const nextIndex = state.currentOptimizedIndex + 1;
        
        if (nextIndex < state.optimizedSequence.length) {
          set(
            { currentOptimizedIndex: nextIndex },
            false,
            'advanceOptimizedStep'
          );
          return true; // Successfully advanced
        }
        return false; // End of sequence
      },
      
      resetOptimizedSequence: () => set(
        { currentOptimizedIndex: 0 },
        false,
        'resetOptimizedSequence'
      ),
      
      // V2 Controller sync - allow V2 state machine to update isActive
      setIsActive: (isActive) => set(
        { isActive },
        false,
        'setIsActive'
      ),
      
      // Custom measure range actions
      setCustomRange: (start, end) => {
        // Store-level validation to prevent invalid state
        if (start < 1 || end < start) {
          console.warn('Invalid range provided to setCustomRange:', { start, end });
          return;
        }
        console.log('[MeasureRangeDebug] setCustomRange called:', { start, end });
        set(
          { 
            customStartMeasure: start, 
            customEndMeasure: end 
          },
          false,
          'setCustomRange'
        );
      },
      
      toggleCustomRange: () => {
        console.log('[MeasureRangeDebug] toggleCustomRange called');
        set(
          (state) => {
            const newActive = !state.customRangeActive;
            console.log('[MeasureRangeDebug] customRangeActive changing:', { 
              from: state.customRangeActive, 
              to: newActive,
              range: { start: state.customStartMeasure, end: state.customEndMeasure }
            });
            return { customRangeActive: newActive };
          },
          false,
          'toggleCustomRange'
        );
      },
      
      clearCustomRange: () => set(
        {
          customRangeActive: false,
          customStartMeasure: 1,
          customEndMeasure: 1
        },
        false,
        'clearCustomRange'
      ),
      
      setCurrentMeasure: (measureIndex) => set(
        { currentMeasureIndex: measureIndex },
        false,
        'setCurrentMeasure'
      ),
      
      setLastValidMeasureIndex: (measureIndex) => set(
        { lastValidMeasureIndex: measureIndex },
        false,
        'setLastValidMeasureIndex'
      )
    }),
    {
      name: 'practice-store',
      trace: process.env.NODE_ENV === 'development'
    }
  )
);