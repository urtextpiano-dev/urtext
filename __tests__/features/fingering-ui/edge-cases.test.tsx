// Additional edge case tests based on AI validation feedback
// These tests address critical scenarios identified by Gemini and Grok3

import { describe, test, expect, beforeEach, jest, afterEach } from '@jest/globals';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import '@testing-library/jest-dom';

// Mock stores and components
jest.mock('@/renderer/features/fingering/hooks/useFingeringStore');
jest.mock('@/renderer/features/practice-mode/stores/practiceStore');

describe('Edge Cases: State Persistence and Race Conditions', () => {
  let mockFingeringStore: any;
  let mockPracticeStore: any;

  beforeEach(() => {
    jest.clearAllMocks();
    
    mockFingeringStore = {
      isEditingMode: false,
      setEditingMode: jest.fn(),
      isEnabled: true
    };
    
    mockPracticeStore = {
      isActive: false,
      setActive: jest.fn()
    };

    require('@/renderer/features/fingering/hooks/useFingeringStore').useFingeringStore = 
      jest.fn(() => mockFingeringStore);
    require('@/renderer/features/practice-mode/stores/practiceStore').usePracticeStore = 
      jest.fn(() => mockPracticeStore);
  });

  describe('Initial State Hydration (Gemini Feedback)', () => {
    test('should correctly show edit mode on initial render when hydrated from storage', () => {
      // Simulate state hydrated from localStorage
      mockFingeringStore.isEditingMode = true; // User had edit mode on when they left
      
      // Mock components that should reflect this state
      const FingeringEditIndicator = () => {
        const { isEditingMode } = require('@/renderer/features/fingering/hooks/useFingeringStore').useFingeringStore();
        return isEditingMode ? <div>Edit Mode</div> : null;
      };
      
      const TopControlsMenu = () => {
        const { isEditingMode } = require('@/renderer/features/fingering/hooks/useFingeringStore').useFingeringStore();
        return (
          <input
            type="checkbox"
            aria-label="Toggle fingering edit mode"
            checked={isEditingMode}
            readOnly
          />
        );
      };
      
      // Render both components
      const { container } = render(
        <>
          <TopControlsMenu />
          <FingeringEditIndicator />
        </>
      );
      
      // Both should immediately reflect the hydrated state
      expect(screen.getByLabelText('Toggle fingering edit mode')).toBeChecked();
      expect(screen.getByText('Edit Mode')).toBeInTheDocument();
      
      // No flicker - should be correct on first render
      expect(container).toMatchSnapshot();
    });

    test('should handle missing/corrupt localStorage gracefully', () => {
      // Simulate corrupt localStorage
      const originalGetItem = Storage.prototype.getItem;
      Storage.prototype.getItem = jest.fn(() => {
        throw new Error('Corrupt storage');
      });
      
      // Component should fall back to defaults
      const TopControlsMenu = () => {
        const { isEditingMode } = require('@/renderer/features/fingering/hooks/useFingeringStore').useFingeringStore();
        return <div>Edit Mode: {isEditingMode ? 'On' : 'Off'}</div>;
      };
      
      render(<TopControlsMenu />);
      
      expect(screen.getByText('Edit Mode: Off')).toBeInTheDocument();
      
      // Restore
      Storage.prototype.getItem = originalGetItem;
    });
  });

  describe('Practice Mode State Transitions (Gemini Feedback)', () => {
    test('should force disable edit mode when practice mode starts', async () => {
      mockFingeringStore.isEditingMode = true; // User is editing
      
      // Mock component that handles the interaction
      const TestComponent = () => {
        const fingeringStore = require('@/renderer/features/fingering/hooks/useFingeringStore').useFingeringStore();
        const practiceStore = require('@/renderer/features/practice-mode/stores/practiceStore').usePracticeStore();
        
        // This logic should be in the actual implementation
        React.useEffect(() => {
          if (practiceStore.isActive && fingeringStore.isEditingMode) {
            fingeringStore.setEditingMode(false);
          }
        }, [practiceStore.isActive]);
        
        return (
          <div>
            <div>Edit Mode: {fingeringStore.isEditingMode ? 'On' : 'Off'}</div>
            <button onClick={() => practiceStore.setActive(true)}>Start Practice</button>
          </div>
        );
      };
      
      const user = userEvent.setup();
      render(<TestComponent />);
      
      expect(screen.getByText('Edit Mode: On')).toBeInTheDocument();
      
      // Start practice mode
      await user.click(screen.getByText('Start Practice'));
      mockPracticeStore.isActive = true;
      
      // Edit mode should be forced off
      expect(mockFingeringStore.setEditingMode).toHaveBeenCalledWith(false);
    });

    test('should not auto-enable edit mode when practice ends', () => {
      // User had edit mode on, practice forced it off
      mockFingeringStore.isEditingMode = false;
      mockPracticeStore.isActive = true;
      
      const TestComponent = () => {
        const fingeringStore = require('@/renderer/features/fingering/hooks/useFingeringStore').useFingeringStore();
        return <div>Edit Mode: {fingeringStore.isEditingMode ? 'On' : 'Off'}</div>;
      };
      
      const { rerender } = render(<TestComponent />);
      
      // End practice mode
      mockPracticeStore.isActive = false;
      rerender(<TestComponent />);
      
      // Edit mode should remain off (not auto-enabled)
      expect(screen.getByText('Edit Mode: Off')).toBeInTheDocument();
      expect(mockFingeringStore.setEditingMode).not.toHaveBeenCalled();
    });
  });

  describe('Rapid Toggle Race Conditions (Grok3 Feedback)', () => {
    test('should debounce rapid edit mode toggles', async () => {
      const user = userEvent.setup();
      
      // Mock component with debounced toggle
      const TopControlsMenu = () => {
        const { isEditingMode, setEditingMode } = require('@/renderer/features/fingering/hooks/useFingeringStore').useFingeringStore();
        const [isProcessing, setIsProcessing] = React.useState(false);
        
        const handleToggle = async (checked: boolean) => {
          if (isProcessing) return; // Prevent rapid clicks
          
          setIsProcessing(true);
          setEditingMode(checked);
          
          // Simulate debounce
          setTimeout(() => setIsProcessing(false), 300);
        };
        
        return (
          <input
            type="checkbox"
            aria-label="Toggle fingering edit mode"
            checked={isEditingMode}
            onChange={(e) => handleToggle(e.target.checked)}
            disabled={isProcessing}
          />
        );
      };
      
      render(<TopControlsMenu />);
      const toggle = screen.getByLabelText('Toggle fingering edit mode');
      
      // Rapid clicks
      await user.click(toggle);
      await user.click(toggle);
      await user.click(toggle);
      
      // Should only process first click due to debounce
      expect(mockFingeringStore.setEditingMode).toHaveBeenCalledTimes(1);
      
      // Toggle should be disabled during processing
      expect(toggle).toBeDisabled();
      
      // Wait for debounce
      await waitFor(() => expect(toggle).not.toBeDisabled(), { timeout: 400 });
    });
  });

  describe('Memory Leak Prevention (Grok3 Feedback)', () => {
    test('should clean up store subscriptions on unmount', () => {
      const unsubscribe = jest.fn();
      const subscribe = jest.fn(() => unsubscribe);
      
      // Mock Zustand subscription
      require('@/renderer/features/fingering/hooks/useFingeringStore').useFingeringStore.subscribe = subscribe;
      
      const TestComponent = () => {
        const store = require('@/renderer/features/fingering/hooks/useFingeringStore').useFingeringStore();
        
        React.useEffect(() => {
          const unsub = store.subscribe(() => {});
          return () => unsub();
        }, []);
        
        return <div>Component</div>;
      };
      
      const { unmount } = render(<TestComponent />);
      
      expect(subscribe).toHaveBeenCalled();
      
      // Unmount should clean up
      unmount();
      expect(unsubscribe).toHaveBeenCalled();
    });
  });

  describe('Unsaved Changes Protection (Grok3 Feedback)', () => {
    test('should warn user about unsaved changes when switching modes', async () => {
      const user = userEvent.setup();
      mockFingeringStore.hasUnsavedChanges = true;
      
      // Mock component with unsaved changes handling
      const TestComponent = () => {
        const handlePracticeStart = () => {
          if (mockFingeringStore.hasUnsavedChanges) {
            if (!window.confirm('You have unsaved fingering changes. Continue?')) {
              return;
            }
          }
          mockPracticeStore.setActive(true);
        };
        
        return (
          <button onClick={handlePracticeStart}>Start Practice</button>
        );
      };
      
      // Mock window.confirm
      window.confirm = jest.fn(() => false);
      
      render(<TestComponent />);
      await user.click(screen.getByText('Start Practice'));
      
      // Should show confirmation
      expect(window.confirm).toHaveBeenCalledWith('You have unsaved fingering changes. Continue?');
      
      // Practice should not start (user cancelled)
      expect(mockPracticeStore.setActive).not.toHaveBeenCalled();
    });
  });
});