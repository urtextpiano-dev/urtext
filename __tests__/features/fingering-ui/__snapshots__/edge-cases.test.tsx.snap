// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Edge Cases: State Persistence and Race Conditions Initial State Hydration (Gemini Feedback) should correctly show edit mode on initial render when hydrated from storage 1`] = `
<div>
  <input
    aria-label="Toggle fingering edit mode"
    checked=""
    readonly=""
    type="checkbox"
  />
  <div>
    Edit Mode
  </div>
</div>
`;
